test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java

```diff
@@ -598,6 +605,30 @@ private static String groupName(ThreadGroup threadGroup) {
         return tempList.subList(0, size);
     }
 
+    public static XContentBuilder shuffleXContent(XContentBuilder builder, Set<String> exceptFieldNames) throws IOException {
+        BytesReference bytes = builder.bytes();
+        XContentParser parser = XContentFactory.xContent(bytes).createParser(bytes);
+        // use ordered maps for reproducibility
+        Map<String, Object> shuffledMap = shuffleMap(parser.mapOrdered(), exceptFieldNames, random());
+        XContentBuilder jsonBuilder = XContentFactory.jsonBuilder();
```

Comment:
it should rather save the XContentType detected from the bytes (XContentFactory.xContent(bytes)) above and re-use it.